cpp/Documentation.md

```diff
@@ -0,0 +1,474 @@
+# C++ Project Documentation
+
+This document provides instructions and hints to document C++ projects.
+
+## How to document code?
+
+### C++ API
+
+Public APIs may be documented with Doxygen, with comments specified in
+headers so that they remain visible to the developers consuming your APIs.
+
+### C++ Internal Code
+
+Documentation of internal code is highly recommended to help the maintainers
+of the project. Using Doxygen is not mandatory though. But if used, then the
+documentations of public and internal code should be well separated, such
+that it remains easy to the reader to distinguish public symbols from internal
+ones.
+
+Documentation of internal symbols can be specified in cpp files where symbols
+are defined, presumably more accessible to the maintainers.
+See [the following discussion](https://stackoverflow.com/questions/355619/where-to-put-the-doxygen-comment-blocks-for-an-internal-library-in-h-or-in-cpp)
+on stackoverflow.
+
+#### Documentation style
+
+Doxygen supports [multiple syntax](http://www.doxygen.nl/manual/docblocks.html)
+to document the code. There is no preferred syntax but Doxygen comments of a particular
+project may use the same one.
+
+Nevertheless, it is suggested to turn the
+[JAVADOC_AUTOBRIEF](http://www.doxygen.nl/manual/config.html#cfg_javadoc_autobrief)
+option to `YES` to shrink comments size:
+
+```c++
+/// A brief function description.
+/// A more elaborated function description.
+void a_function();
+```
+
+instead of:
+```c++
+/// \brief A brief function description.
+///
+/// A more elaborated function description.
+void a_function();
+
+
+#### Document files
+
+Document header files, and source files if possible by adding a `\file` paragraph in the
+file prelude.
+
+```c++
+/// \file
+/// A brief file description
+/// A more elaborated file description
+```
+
+#### Only document the declarations
+
+Only function or class declarations should be documented, not the definition.
+
+```c++
+/// Helper function to create a vertex unique identifier.
+/// \param type vertex type
+/// \param id vertex identifier
+vertex_uid_t make_id(vertex_t type, vertex_id_t id);
+
+// [...]
+
```

Comment:
see my comments above, I think that indeed this example here is correct, however if there is more than just the `\brief`, the `\param` and `\return` that should go to the defintion (and into the `.cpp` file).

Reply:
I updated the example in 53a76c9